File: framework/languages/java/samples/java/DeliveryDispatch/Server/DispatchApi/src/main/java/systems/zlink/samples/deliverydispatch/server/dispatchapi/handlers/ServerAssertionHandler.java
package systems.zlink.samples.deliverydispatch.server.dispatchapi.handlers;

import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;
import systems.zlink.samples.deliverydispatch.server.configuration.EvidenceStore;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages.Status;

@RestController
public final class ServerAssertionHandler {
    private final EvidenceStore evidence;

    public ServerAssertionHandler(EvidenceStore evidence) {
        this.evidence = evidence;
    }

    @PostMapping("/self-check/assert")
    public Messages.ServerAssertionRes handle(@RequestBody Messages.ServerAssertionReq request) {
        boolean success = evidence.hasSequence(
            request.successfulDeliveryId(),
            Status.Assigned,
            Status.Accepted,
            Status.PickedUp,
            Status.Delivered);
        boolean reassigned = evidence.hasSequence(
            request.reassignedDeliveryId(),
            Status.Assigned,
            Status.Reassigned,
            Status.Accepted,
            Status.PickedUp,
            Status.Delivered);
        return new Messages.ServerAssertionRes(success && reassigned, evidence.readLines());
    }
}

File: framework/languages/java/samples/kotlin/DeliveryDispatch/Server/DispatchApi/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/dispatchapi/handlers/ServerAssertionHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.dispatchapi.handlers

import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.EvidenceStore
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatuses
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.ServerAssertionReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.ServerAssertionRes

@RestController
class ServerAssertionHandler(
    private val evidence: EvidenceStore,
) {
    @PostMapping("/self-check/assert")
    fun handle(@RequestBody request: ServerAssertionReq): ServerAssertionRes {
        val success = evidence.hasSequence(
            request.successfulDeliveryId,
            DeliveryStatuses.Assigned,
            DeliveryStatuses.Accepted,
            DeliveryStatuses.PickedUp,
            DeliveryStatuses.Delivered,
        )
        val reassigned = evidence.hasSequence(
            request.reassignedDeliveryId,
            DeliveryStatuses.Assigned,
            DeliveryStatuses.Reassigned,
            DeliveryStatuses.Accepted,
            DeliveryStatuses.PickedUp,
            DeliveryStatuses.Delivered,
        )
        return ServerAssertionRes(success && reassigned, evidence.readLines())
    }
}
